modules/src/clients/ics07_tendermint/client_def.rs

```diff
@@ -185,116 +190,278 @@ impl ClientDef for TendermintClient {
         &self,
-        _client_state: &Self::ClientState,
-        _height: Height,
-        _proof: &CommitmentProofBytes,
-        _port_id: &PortId,
-        _channel_id: &ChannelId,
-        _seq: &Sequence,
+        ctx: &dyn ChannelReader,
+        client_state: &Self::ClientState,
+        height: Height,
+        connection_end: &ConnectionEnd,
+        proof: &CommitmentProofBytes,
+        root: &CommitmentRoot,
+        port_id: &PortId,
+        channel_id: &ChannelId,
+        sequence: Sequence,
     ) -> Result<(), Ics02Error> {
-        todo!()
+        client_state.verify_height(height)?;
+        verify_delay_passed(ctx, height, connection_end)?;
+
+        let seq_path = Path::SeqRecvs(port_id.clone(), channel_id.clone());
+        verify_membership(
+            client_state,
+            connection_end.counterparty().prefix(),
+            proof,
+            root,
+            seq_path.to_string(),
+            u64::from(sequence).encode_to_vec(),
+        )
     }
 
     fn verify_packet_receipt_absence(
         &self,
-        _client_state: &Self::ClientState,
-        _height: Height,
-        _proof: &CommitmentProofBytes,
-        _port_id: &PortId,
-        _channel_id: &ChannelId,
-        _seq: &Sequence,
+        ctx: &dyn ChannelReader,
+        client_state: &Self::ClientState,
+        height: Height,
+        connection_end: &ConnectionEnd,
+        proof: &CommitmentProofBytes,
+        root: &CommitmentRoot,
+        port_id: &PortId,
+        channel_id: &ChannelId,
+        sequence: Sequence,
     ) -> Result<(), Ics02Error> {
-        todo!()
+        client_state.verify_height(height)?;
+        verify_delay_passed(ctx, height, connection_end)?;
+
+        let receipt_path = Path::Receipts {
+            port_id: port_id.clone(),
+            channel_id: channel_id.clone(),
+            sequence,
+        };
+        verify_non_membership(
+            client_state,
+            connection_end.counterparty().prefix(),
+            proof,
+            root,
+            receipt_path.to_string(),
+        )
     }
 
     fn verify_upgrade_and_update_state(
         &self,
         _client_state: &Self::ClientState,
         _consensus_state: &Self::ConsensusState,
-        _proof_upgrade_client: MerkleProof,
-        _proof_upgrade_consensus_state: MerkleProof,
+        _proof_upgrade_client: RawMerkleProof,
+        _proof_upgrade_consensus_state: RawMerkleProof,
     ) -> Result<(Self::ClientState, Self::ConsensusState), Ics02Error> {
         todo!()
     }
 }
 
+fn verify_membership(
+    client_state: &ClientState,
+    prefix: &CommitmentPrefix,
+    proof: &CommitmentProofBytes,
+    root: &CommitmentRoot,
+    path: String,
+    value: Vec<u8>,
+) -> Result<(), Ics02Error> {
+    let merkle_path = apply_prefix(prefix, vec![path]).map_err(Error::ics23_error)?;
+    let merkle_proof: MerkleProof = RawMerkleProof::try_from(proof.clone())
+        .map_err(Ics02Error::invalid_consensus_state_proof)?
```

Comment:
Not sure if the error type `invalid_consensus_state_proof` is right here. The proof is a _not_ for a consensus state, but can be for a connection state, channel state, client state, ack, packet, etc. It's actually a general commitment proof. We might need to defined a new proof [variant in `Ics02Error` to use](https://github.com/informalsystems/ibc-rs/blob/aa2336bdd9ae5cb210c61434dc12b8be58f94be7/modules/src/core/ics02_client/error.rs#L19) here, for example:

```rust
        InvalidCommitmentProof
            [ Ics23Error ]
            | _ | { "invalid commitment proof bytes" },
```

Reply:
11c641df5a34a74d1228f7dee7643648034a9934